File: blue-core/src/blue/blueLive/LiveObject.java
/*
 * blue - object composition environment for csound
 * Copyright (c) 2000-2006 Steven Yi (stevenyi@gmail.com)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published
 * by  the Free Software Foundation; either version 2 of the License or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.LIB.  If not, write to
 * the Free Software Foundation Inc., 59 Temple Place - Suite 330,
 * Boston, MA  02111-1307 USA
 */

package blue.blueLive;

import blue.SoundObjectLibrary;
import blue.soundObject.SoundObject;
import blue.utility.ObjectUtilities;
import blue.utility.XMLUtilities;
import electric.xml.Element;
import electric.xml.Elements;
import java.io.Serializable;
import java.rmi.dgc.VMID;

/**
 * 
 * @author steven
 */
public class LiveObject implements Serializable {

    private SoundObject sObj = null;

    private int midiTrigger = -1;

    private int keyTrigger = -1;
    
    private boolean enabled = false;
    
    private String uniqueId;

    /** Creates a new instance of LiveObject */
    public LiveObject() {
        uniqueId = Integer.toString(new VMID().hashCode());
    }

    public LiveObject(SoundObject sObj) {
        this();
        this.sObj = sObj;
    }

    public SoundObject getSoundObject() {
        return sObj;
    }

    public void setSObj(SoundObject sObj) {
        this.sObj = sObj;
    }

    public int getMidiTrigger() {
        return midiTrigger;
    }

    public void setMidiTrigger(int midiTrigger) {
        this.midiTrigger = midiTrigger;
    }

    public int getKeyTrigger() {
        return keyTrigger;
    }

    public void setKeyTrigger(int keyTrigger) {
        this.keyTrigger = keyTrigger;
    }

    public boolean isEnabled() {
        return enabled;
    }

    public void setEnabled(boolean enabled) {
        this.enabled = enabled;
    }
    
    public String getUniqueId() {
        return uniqueId;
    }

    public Element saveAsXML(SoundObjectLibrary sObjLibrary) {
        Element retVal = new Element("liveObject");
        
        retVal.setAttribute("uniqueId", uniqueId);

        retVal.addElement(XMLUtilities.writeInt("keyTrigger", keyTrigger));
        retVal.addElement(XMLUtilities.writeInt("midiTrigger", midiTrigger));
        retVal.addElement(XMLUtilities.writeBoolean("enabled", enabled));
        retVal.addElement(this.sObj.saveAsXML(sObjLibrary));

        return retVal;
    }

    public static LiveObject loadFromXML(Element data,
            SoundObjectLibrary sObjLibrary) throws Exception {
        LiveObject liveObj = new LiveObject();
        
        String val = data.getAttributeValue("uniqueId");
        if (val != null && val.length() > 0) {
            liveObj.uniqueId = val;
        }

        Elements nodes = data.getElements();

        while (nodes.hasMoreElements()) {
            Element node = nodes.next();
            String name = node.getName();

            if (name.equals("keyTrigger")) {
                liveObj.setKeyTrigger(XMLUtilities.readInt(node));
            } else if (name.equals("midiTrigger")) {
                liveObj.setMidiTrigger(XMLUtilities.readInt(node));
            } else if (name.equals("soundObject")) {
                liveObj.setSObj((SoundObject) ObjectUtilities.loadFromXML(node,
                        sObjLibrary));
            } else if (name.equals("enabled")) {
                liveObj.setEnabled(XMLUtilities.readBoolean(node));
            }
        }

        return liveObj;
    }
}

File: packages/blue-data/src/live/live-object.ts
import { Element } from '../serialization/xml-reader';
import { ObjRefSaveMap, ObjRefLoadMap } from '../serialization/obj-ref-map';
import { BlueDataObject } from '../blue-data-object';
import { readInt, writeInt, readBoolean, writeBoolean } from '../utilities/xml';
import { loadSoundObjectFromXML } from '../sound-objects/sound-object-registry';
import '../sound-objects/register-sound-object-types';
import type { SoundObject } from '../sound-objects/sound-object';

function generateUniqueId(): string {
  return `${Date.now()}-${Math.floor(Math.random() * 2147483647)}`;
}

export class LiveObject implements BlueDataObject {
  private _uniqueId: string;
  private _keyTrigger = -1;
  private _midiTrigger = -1;
  private _enabled = false;
  private _soundObject: SoundObject | null = null;

  constructor() {
    this._uniqueId = generateUniqueId();
  }

  getUniqueId(): string {
    return this._uniqueId;
  }

  setUniqueId(id: string): void {
    this._uniqueId = id;
  }

  getKeyTrigger(): number {
    return this._keyTrigger;
  }

  setKeyTrigger(v: number): void {
    this._keyTrigger = v;
  }

  getMidiTrigger(): number {
    return this._midiTrigger;
  }

  setMidiTrigger(v: number): void {
    this._midiTrigger = v;
  }

  isEnabled(): boolean {
    return this._enabled;
  }

  setEnabled(v: boolean): void {
    this._enabled = v;
  }

  getSoundObject(): SoundObject | null {
    return this._soundObject;
  }

  setSoundObject(sObj: SoundObject | null): void {
    this._soundObject = sObj;
  }

  getDisplayName(): string {
    return this._soundObject?.getName() ?? '';
  }

  getSoundObjectType(): string {
    if (!this._soundObject) return '';
    return (
      ((this._soundObject as unknown as Record<string, unknown>).constructor?.name as string) ?? ''
    );
  }

  get hasSoundObject(): boolean {
    return this._soundObject !== null;
  }

  saveAsXML(objRefMap?: ObjRefSaveMap): Element {
    const elem = new Element('liveObject');
    elem.setAttribute('uniqueId', this._uniqueId);
    elem.addElement(writeInt('keyTrigger', this._keyTrigger));
    elem.addElement(writeInt('midiTrigger', this._midiTrigger));
    elem.addElement(writeBoolean('enabled', this._enabled));
    if (this._soundObject) {
      elem.addElement(this._soundObject.saveAsXML(objRefMap));
    }
    return elem;
  }

  static loadFromXML(data: Element, objRefMap?: ObjRefLoadMap): LiveObject {
    const obj = new LiveObject();
    const uniqueId = data.getAttribute('uniqueId');
    if (uniqueId) {
      obj._uniqueId = uniqueId;
    }
    const nodes = data.getElements();
    while (nodes.hasMoreElements()) {
      const node = nodes.next();
      const name = node.getName();
      switch (name) {
        case 'keyTrigger':
          obj._keyTrigger = readInt(node);
          break;
        case 'midiTrigger':
          obj._midiTrigger = readInt(node);
          break;
        case 'enabled':
          obj._enabled = readBoolean(node);
          break;
        case 'soundObject': {
          const sObj = loadSoundObjectFromXML(node, objRefMap);
          if (sObj) obj._soundObject = sObj;
          break;
        }
      }
    }
    return obj;
  }

  deepCopy(): BlueDataObject {
    const copy = new LiveObject();
    copy._uniqueId = this._uniqueId;
    copy._keyTrigger = this._keyTrigger;
    copy._midiTrigger = this._midiTrigger;
    copy._enabled = this._enabled;
    if (this._soundObject) {
      copy._soundObject = this._soundObject.deepCopy() as SoundObject;
    }
    return copy;
  }
}
